x-element.js

```diff
@@ -1710,39 +1698,146 @@ class TemplateEngine {
+          break;
+        default:
+          TemplateEngine.#throwUpdaterError(context.updater, 'content');
+          break;
+      }
+    } else {
+      TemplateEngine.#resetIfRequired(node, startNode, value, lastValue, context, lastContext);
+      TemplateEngine.#content(node, startNode , value, lastValue);
+    }
+  }
+
+  static #commitText(node, value, lastValue) {
+    const context = TemplateEngine.#updaterMap.get(value);
+    if (context) {
+      TemplateEngine.#throwUpdaterError(context.updater, 'text');
+    } else {
+      TemplateEngine.#text(node, value, lastValue);
+    }
+  }
+
+  static #commit(result) {
+    const { directions, values, lastValues } = result;
+    for (const { key, type, node, startNode, name } of directions) {
+      const value = values[key];
+      const lastValue = lastValues[key];
+      switch (type) {
+        case 'attribute': TemplateEngine.#commitAttribute(node, name, value, lastValue); break;
+        case 'boolean': TemplateEngine.#commitBoolean(node, name, value, lastValue); break;
+        case 'defined': TemplateEngine.#commitDefined(node, name, value, lastValue); break;
+        case 'property':TemplateEngine.#commitProperty(node, name, value, lastValue); break;
+        case 'content': TemplateEngine.#commitContent(node, startNode, value, lastValue); break;
+        case 'text': TemplateEngine.#commitText(node, value, lastValue); break;
+      }
+    }
+  }
+
+  static #throwUpdaterError(updater, type) {
+    switch (updater) {
+      case TemplateEngine.#live:
+        throw new Error(`The live update must be used on ${TemplateEngine.#getTypeText('property')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+      case TemplateEngine.#map:
+        throw new Error(`The map update must be used on ${TemplateEngine.#getTypeText('content')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+      case TemplateEngine.#unsafeHTML:
+        throw new Error(`The unsafeHTML update must be used on ${TemplateEngine.#getTypeText('content')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+      case TemplateEngine.#unsafeSVG:
+        throw new Error(`The unsafeSVG update must be used on ${TemplateEngine.#getTypeText('content')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+
+      // We’ll delete these updaters later.
+      case TemplateEngine.#ifDefined:
+        throw new Error(`The ifDefined update must be used on ${TemplateEngine.#getTypeText('attribute')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+      case TemplateEngine.#nullish:
+        throw new Error(`The nullish update must be used on ${TemplateEngine.#getTypeText('attribute')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+      case TemplateEngine.#repeat:
+        throw new Error(`The repeat update must be used on ${TemplateEngine.#getTypeText('content')}, not on ${TemplateEngine.#getTypeText(type)}.`);
+    }
+  }
+
+  static #resetIfRequired(node, startNode, value, lastValue, context, lastContext) {
```

Comment:
In case you’re curious — this function fixes a _bunch_ of nuanced edge cases surrounding the transition from binding one kind of content value to another.